csharp_dotnetcore/14.NLP-With-Dispatch/NlpDispatch/NlpDispatchBot.cs

```diff
@@ -0,0 +1,182 @@
+        /// </summary>
+        private async Task DispatchToTopIntentAsync(ITurnContext context, (string intent, double score)? topIntent)
+        {
+            const string homeAutomationDispatchKey = "l_homeautomation-LUIS";
+            const string weatherDispatchKey = "l_weather-LUIS";
+            const string noneDispatchKey = "None";
+            const string qnaDispatchKey = "q_sample_qna";
+
+            switch (topIntent.Value.intent.ToLowerInvariant())
+            {
+                case homeAutomationDispatchKey:
+                    await DispatchToLuisModelAsync(context, HomeAutomationLuisKey);
+
+                    // Here, you can add code for calling the hypothetical home automation service, passing in any entity information that you need
+                    break;
+                case weatherDispatchKey:
+                    await DispatchToLuisModelAsync(context, WeatherLuisKey);
+
+                    // Here, you can add code for calling the hypothetical weather service,
+                    // passing in any entity information that you need
+                    break;
+                case noneDispatchKey:
+                    // You can provide logic here to handle the known None intent (none of the above).
+                    // In this example we fall through to the QnA intent.
+                case qnaDispatchKey:
+                    await DispatchToQnAMakerAsync(context, QnAMakerKey);
+                    break;
+
+                default:
+                    // The intent didn't match any case, so just display the recognition results.
+                    await context.SendActivityAsync($"Dispatch intent: {topIntent.Value.intent} ({topIntent.Value.score}).");
+
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Dispatches the turn to the request QnaMaker app.
+        /// </summary>
+        private async Task DispatchToQnAMakerAsync(ITurnContext context, string appName)
+        {
+            if (!string.IsNullOrEmpty(context.Activity.Text))
+            {
+                var results = await _services.QnAServices[appName].GetAnswersAsync(context).ConfigureAwait(false);
+                if (results.Any())
+                {
+                    await context.SendActivityAsync(results.First().Answer);
+                }
+                else
+                {
+                    await context.SendActivityAsync($"Couldn't find an answer in the {appName}.");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Dispatches the turn to the requested Luis model.
+        /// </summary>
+        private async Task DispatchToLuisModelAsync(ITurnContext context, string appName)
+        {
+            await context.SendActivityAsync($"Sending your request to the {appName} system ...");
+            var result = await _services.LuisServices[appName].RecognizeAsync(context, CancellationToken.None);
```

Comment:
>CancellationToken.N [](start = 87, length = 19)

Hrmmmmmmm...

Reply:
hahahaha I completely forgot about this 🥇 No comment!